doc/compiler/runtime/CodeMetaData.md

```diff
@@ -0,0 +1,88 @@
+code metadata and code metadata manager in the Eclipse OMR compiler.
+
+## Code Metadata
+
+Code metadata contains runtime metadata information about a compilation unit. 
+Metadata may include start address of allocated code memory and its size, 
+start PC address of compiled code (when it is invoked from an interpreter 
+or invoked from another piece of compiled code), end PC address where the 
+compiled code ends, compilation hotness level, list of runtime assumptions 
+made during the compilation and any other information the compiler may find 
+useful during runtime.
+
+An instance of code metadata will be created for each compilation unit after
+it is compiled. The code metadata instance will exist for the duration of the 
+compilation unit's runtime until it is unloaded or invalidated. Storage and 
+look up of code metadata instances are carried out by the code metadata manager.
+
+In the Eclipse OMR compiler, an instance of code metadata is uniquely identified 
+by its start and end PC addresses. Two instances of code metadata with PC address 
+ranges that overlap are not allowed.
+
+## Code Metadata Manager
+
+Code metadata manager is a singleton class created in the runtime to manage 
+code metadata instances. The manager class constructor allocates an AVL tree 
+in persistent memory to facilitate storage and look up of code metadata instances 
+created by the compiler. Each tree node in the AVL tree is a hash table representing 
+a piece of code cache segment. A node is created and added to the AVL tree when a 
+new code cache segment is allocated. For [Eclipse OpenJ9](https://github.com/eclipse/openj9), 
+code cache segment size is configured to be 2MB. Entries in each hash table correspond 
+to equal chunks of the associated code cache. For [Eclipse OpenJ9](https://github.com/eclipse/openj9), 
+each hash table has 4096 entries and each entry is responsible for 512 bytes of the 2MB 
+code cache. A hash table entry points to a metadata pointer or a list of metadata 
+pointers whose compiled code occupies a part of the chunk or the entire chunk of 
+the code cache.
+
+For efficient insertion, deletion, and look-up of a particular metadata pointer, 
+please note that to add or remove nodes from the AVL tree, we will need exclusive 
+VM access, which means threads with VM access have to be stopped before the operation 
+can be carried out. However, to add an entry to a hash table or to look up a metadata 
+pointer given a PC address, we only need VM access, which means the operation can be 
+carried out concurrently with other threads holding VM access. Removing a metadata 
+pointer from a hash table in [Eclipse OpenJ9](https://github.com/eclipse/openj9) is 
+done at the end of GC when a method is unloaded or recompiled and reclaimed, 
+we already have exclusive VM access at that point.
+
+When a metadata pointer is added with the start and end PC addresses of its 
+compiled code, we need to do the following: 
```

Comment:
The following is the procedure for adding a metadata pointer with the start and end PC addresses of its compiled code: